secretsmanager/src/SecretRetriever.cpp

```diff
@@ -0,0 +1,91 @@
+// Concord
+//
+// Copyright (c) 2021 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License"). You may not use
+// this product except in compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright notices and license
+// terms. Your use of these subcomponents is subject to the terms and conditions of the
+// subcomponent's license, as noted in the LICENSE file.
+
+#include "SecretRetriever.hpp"
+#include <iostream>
+#include <iomanip>
+#include <fstream>
+#include <exception>
+#include <optional>
+#include <regex>
+#include <nlohmann/json.hpp>
+#include "httplib.h"
+
+using json = nlohmann::json;
+using namespace std::string_literals;
+
+namespace SecretRetriever {
+
+SecretData ParseJson(json j) {
+  return SecretData{j["password"].get<std::string>(),
+                    j["message_digest_algorithm"].get<std::string>(),
+                    j["encryption_algorithm"].get<std::string>(),
+                    (uint32_t)std::stoi(j["key_length"].get<std::string>())};
+}
+
+std::optional<SecretData> FileSecretRetrieve(const std::string &secret_url) {
+  std::regex url_regex(R"(^file://(.*))", std::regex::extended);
+  std::smatch url_match;
+  if (std::regex_match(secret_url, url_match, url_regex)) {
+    std::ifstream jsonFile(url_match[1], std::ifstream::in);
+    if (jsonFile.fail()) throw std::runtime_error("Secret retrieve failed to open "s + secret_url);
+    json j;
+    jsonFile >> j;
+    return ParseJson(j);
+  } else {
+    return std::nullopt;
+  }
+}
+
+std::optional<SecretData> RemoteSecretRetrieve(const std::string &secret_url) {
+  std::regex url_regex(R"(^(https|http)://([a-zA-Z0-9\.-]+)(:([0-9]+))?(.*))", std::regex::extended);
```

Comment:
We decided to allow http for two main reasons:

1. We have got a mock for the Secrets API which currently doesn't support HTTPS. We want to add some apollo tests and this will make our life easier.
2. The Secrets API in production will work on localhost, so there might be a case where someone might want to use HTTP instead of HTTPS.

Or in nutshell - we deliberately decided to let our users shoot theirselves in the foot if they insist.

Reply:
Small update here - Secrets API will listen on HTTP by design, so we need to support it.